### ISSUE TITLE
peerstore shouldn't be modified on disk when peers are set manually

### ISSUE THREAD

**Issue by ItalyPaleAle** [ipfs/ipfs-cluster#648]
#### Basic information

* [x] Version information: 0.7.0+git7e1ddd90a7afa3b1eb35ec3126d4dda83f0be13e
* [x] Type: Bug
* [x] Operating system: Tested on Linux
* [x] Installation method: Docker

#### Description

When I created my cluster, I defined the list of peers manually (the peers communicate over the public Internet, and this was done to reduce the number of ports that need to be open). I put each peer's multiaddress in the `peerstore` file.

The cluster has been running for a couple of months now. However, a couple of times I had an incident in which one peer was down for a prolonged amount of time (over a few hours). In those situations, restarting the node didn't bring back the peer in the cluster automatically. I have found that it was because the other peers had modified the `peerstore` file on disk, removing the peer.

In short, when the cluster is created with a fixed `peerstore` file, ipfs-cluster should never modify it, even if a peer is offline.

(Here's [how I set up my cluster](https://withblue.ink/2018/11/14/distributed-web-host-your-website-with-ipfs-clusters-cloudflare-and-devops.html ), if you need more info)

**Comment by hsanjuan** [comment]
Hi! Really cool blog post!

The only moment the `peerstore` file is saved is during shutdown. The multiaddresses that are saved in it are those of the peers that are currently known in the libp2p Host peerstore. The peer that went down should have been known to the peerstore even if it was down, unless something removed it from it.

I'm starting to think (from this and a similar issue before) that some recent version of libp2p may have started to GC addresses from the peerstore when connections to them fail. I will need to check this and will update this issue. In any case, you are right that your peers should not be removed from the peerstore when they are down.

On the other side, when the peer came back, if it did keep its peerstore intact, it should have been re-discovered via DHT as soon as it connected to someone, even if other peers did not its multiaddress anymore. Are your peers behind NAT?

By the way, since you're here, I saw in your blog post:

> Why not using the method of Starting a single peer and bootstrapping the rest to it? Because
that requires the nodes to communicate on the API port 9094, which should not be exposed 
on the public Internet unless TLS is configured (and doing that would require setting up TLS 
certificates, basic auth, etc).


That's not accurate, bootstrapping is done to 9096 and uses the cluster secret. Let me know if somewhere in the docs we wrote 9094 and this lead you to this conclusion, as we would need to fix it.

**Comment by ItalyPaleAle** [comment]
Thanks for responding quickly!

Yes, I think that pointing the fingers at GC in libp2p makes a lot of sense. Would also explain the issues I've had with the cluster, and why a node brought back online (after a long downtime) sometimes has issues communicating with the other nodes, and that's why I ended up restarting all nodes, one at a time. I also notice in the `peerstore` file a bunch of weird addresses, sometimes pointing to private IPs (which aren't routable from the node), and sometimes including the DNS name of the public IP I had originally set.

All peers have can receive incoming traffic on port 9096, with no firewalls blocking that. But I have seen weird behaviors with peers not being all able to talk to each other if one goes down for a while. I'm happy to debug more next time I see that.

I would suggest investigating a way to turn libp2p into "full on manual mode" whenever the admin wants to do that? So there'd be no GC, and no changes to the `peerstore` file.

PS: With regards to the blog post... Thanks for reading it and for pointing out that mistake. I checked the post and didn't notice it referring to port 9094 for ipfs-cluster anywhere else; I think I just made a typo in that spot. I'll fix it right away :)

**Comment by ItalyPaleAle** [comment]
Oh, ok, so that shouldn't be an issue.

Please let me know if you can find any issue with libp2p here.

I also noticed that you released version 0.8 a couple of days back, so I'll work over the long weekend on updating my cluster and the documentation, and see if the cluster performs any different.

PS: Totally unrelated and happy to open a new issue for that... In the post I wrote that for ipfs-cluster you should always use an odd number of nodes, at least 3. This is because of my understanding of how most distributed consensus algorithms work, and I have found one other person recommending that with ipfs-cluster specifically. However, the official documentation doesn't make recommendations on that front (or, at least, I didn't see that). Do you have an official opinion on that?

**Comment by hsanjuan** [comment]
> Do you have an official opinion on that?

Yeah, this is because, when electing a leader, an even numbers of peers may result in ties more easily and that requires another election round. But from my experience, Raft probably has some safeguards to prevent this from happenning perpetually by randomly who is voted, and leader election does not happen so often.

The main problem is when using just 2 peers. If you use 2 peers and one is down, the other peer cannot elect itself leader because his vote does not reach >50%. But anything 3+ should work fine, even though with odd numbers you have less chances of having to repeat election rounds when they happen.

After a quick look, I cannot find any straight mention to using odd numbers in Consul (which uses the same Raft library), other than a mention saying that Autopilot tries to keep odd voters (https://www.hashicorp.com/blog/consul-0-8), so I guess that yes, it's slightly better, but not essential.

**Comment by ItalyPaleAle** [comment]
My understanding (which is limited :) ) is that the biggest risk with even-numbered clusters is that they could end up in "split brains" conditions.

Anyways, good to know you recommend at least 3 nodes.
